collector/arp_linux.go

```diff
@@ -73,40 +89,54 @@ func getTotalArpEntries(deviceEntries []procfs.ARPEntry) map[string]uint32 {
 	return entries
 }
 
-func getTotalArpEntriesRTNL() (map[string]uint32, error) {
+func getArpEntriesRTNL() (map[string]uint32, map[string]map[string]int, error) {
 	conn, err := rtnl.Dial(nil)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 	defer conn.Close()
 
 	// Neighbors will also contain IPv6 neighbors, but since this is purely an ARP collector,
 	// restrict to AF_INET.
 	neighbors, err := conn.Neighbours(nil, unix.AF_INET)
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
 
 	// Map of interface name to ARP neighbor count.
 	entries := make(map[string]uint32)
+	// Map of map[InterfaceName]map[StateName]int
+	states := make(map[string]map[string]int)
 
 	for _, n := range neighbors {
 		// Skip entries which have state NUD_NOARP to conform to output of /proc/net/arp.
-		if n.State&unix.NUD_NOARP == 0 {
-			entries[n.Interface.Name]++
+		if n.State&unix.NUD_NOARP == unix.NUD_NOARP {
+			continue
+		}
+
+		entries[n.Interface.Name]++
+
+		_, ok := states[n.Interface.Name]
+		if !ok {
+			states[n.Interface.Name] = make(map[string]int)
 		}
+
+		states[n.Interface.Name][neighborStatesMap[n.State]]++
 	}
 
-	return entries, nil
+	return entries, states, nil
 }
 
 func (c *arpCollector) Update(ch chan<- prometheus.Metric) error {
-	var enumeratedEntry map[string]uint32
+	var (
+		enumeratedEntry map[string]uint32
+		enumStates      map[string]map[string]int
```

Comment:
Why use a signed `int` here? Can the state counts be negative?

Reply:
Changed to that, good point. Thanks.